wifiphisher/common/extensions.py

```diff
@@ -0,0 +1,305 @@
+
+    def get_output(self):
+        """
+        Gets the output of each extensions.
+        Merges them in a list and returns it.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        output = []
+        for extension in self._extensions:
+            m_output = extension.send_output()
+            if m_output and len(m_output) > 0:
+                output += m_output
+        return output
+
+    def _process_packet(self, pkt):
+        """
+        Pass each captured packet to each module.
+        Gets the packets to send.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :param pkt: A Scapy packet object
+        :type pkt: Scapy Packet
+        :return: None
+        :rtype: None
+        """
+
+        for extension in self._extensions:
+            channel_nums, received_packets = extension.get_packet(pkt)
+            if received_packets and len(received_packets) > 0:
+                for c_num in channel_nums:
+                    self._packets_to_send[c_num] += received_packets
+
+    def _stopfilter(self, pkt):
+        """
+        A scapy filter to determine if we need to stop.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :param self: A Scapy packet object
+        :type self: Scapy Packet
+        :return: True or False
+        :rtype: Boolean
+        """
+
+        return not self._should_continue
+
+    def _listen(self):
+        """
+        Listening thread. Listens for packets and forwards them
+        to _process_packet.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        try:
+            # continue to find clients until told otherwise
+            while self._should_continue:
+                dot11.sniff(
+                    iface=self._interface.get_name(),
+                    prn=self._process_packet,
+                    count=1,
+                    store=0,
+                    stop_filter=self._stopfilter)
+        # Don't display "Network is down" if shutting down
+        except OSError:
+            if not self._should_continue:
+                pass
+
+    def _send(self):
+        """
+        Sending thread. Continously broadcasting packets
+        crafted by extensions.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        try:
+            while self._should_continue:
+                for pkt in self._packets_to_send[self._current_channel] + \
+                        self._packets_to_send["*"]:
+                    try:
+                        self._socket.send(pkt)
+                    except BaseException:
+                        continue
```

Comment:
and for the following `time.sleep(1)`. since it's may not be straightforward why we want to do a sleep here.